src/pages/FreeClass/FreeClassrooms.tsx

```diff
@@ -0,0 +1,326 @@
+  },
+  { id: 1, type: SearchClassType.HEADQUARTER, text: ["Scegli la tua", "sede"] },
+]
+
+let searchTimeout: NodeJS.Timeout
+const deltaTime = 200 //ms
+
+/**
+ * This is the first page where the user select the modality to find a free room.
+ * There are two options:
+ * - Selection by campus
+ * - Selection by current user position->this feature is available ONLY IF the GPS is enabled.
+ */
+export const FreeClassrooms: MainStackScreen<"FreeClassrooms"> = () => {
+  const [search, setSearch] = useState("")
+  const { navigate } = useNavigation()
+  const { palette } = usePalette()
+
+  const { rooms, setRooms, date, acronym, setIsRoomsSearching } = useContext(
+    RoomsSearchDataContext
+  )
+
+  //rooms filtered locally
+  const [searchableRooms, setSearchableRooms] = useState<Partial<Room>[]>([])
+
+  //shown rooms
+  const [actualSearchableRooms, setActualSearchableRooms] = useState<Room[]>([])
+
+  //for searching in multiple campuses
+  const [acronymList, setAcronymList] = useState<ValidAcronym[]>([])
+
+  //keep track of when to show activity indicator
+  const [isSearchBarSearching, setIsSearchBarSearching] = useState(false)
+
+  //main function that handles the call to the API in order to obtain the list of freeclassRooms
+  const getAllRoomsFromApi = async (
+    overrideAcr?: ValidAcronym,
+    overrideDate?: Date,
+    overrideSearchBehaviour?: boolean
+  ) => {
+    const searchAcronym = overrideAcr ?? acronym
+    const searchDate = overrideDate ?? date
+
+    if (!acronym) {
+      return
+    }
+    //Check if stored rooms are still relevant to the current search
+    const prevSearchDateISO = rooms[searchAcronym].searchDate
+    if (prevSearchDateISO) {
+      const prevSearchDate = new Date(prevSearchDateISO)
+      if (isSameDay(prevSearchDate, searchDate)) {
+        const currentExpirationDateISO = rooms[searchAcronym].expireAt
+        if (currentExpirationDateISO) {
+          const currentExpirationDate = new Date(currentExpirationDateISO)
+          if (new Date() < currentExpirationDate) {
+            // not expired and relevant
+            return
+          }
+        }
+      }
+    }
+    //expired and not relevant
+
+    if (!overrideSearchBehaviour) {
+      setIsRoomsSearching(true)
+    }
+    const { data, expire } = await api.rooms.getFreeRoomsDay(
+      searchAcronym,
+      formatDate(searchDate),
+      { maxRetries: 1, retryType: RetryType.RETRY_N_TIMES }
+    )
+    if (data.length > 0) {
+      const newGlobalRooms = { ...rooms }
+      newGlobalRooms[searchAcronym].rooms = data
+      const expirationDate = getExpirationDateRooms(expire)
+      //update expiration date or reset
+      newGlobalRooms[searchAcronym].expireAt =
+        expirationDate?.toISOString() ?? undefined
+      //update searchDate
+      newGlobalRooms[searchAcronym].searchDate = searchDate.toISOString()
+      setRooms(newGlobalRooms)
+    }
+    if (!overrideSearchBehaviour) {
+      setIsRoomsSearching(false)
+    }
+  }
```

Comment:
Di nuovo, i side-effect sono il male di questo mondo, non chiedetemi il perché ma è così

questa funzione sarebbe meglio renderla pura e farla restituire qualcosa, punti bonus se tutte le variabili interne sono passate come parametri così può essere estratta dal componente, poi si setta lo state con ciò che viene restituito

Reply:
Cosa significa renderla pura?

Reply:
allora in programmazione funzionale una funzione è pura se dipende unicamente dai parametri passati in input e da un output senza alcun side-effect (ovvero in qualche modo causare una modifica di qualsiasi stato o variabile esterno alla funzione stessa), in questo caso ad esempio vengono chiamati gli state setter di React, che non è bellissimo

però in realtà ho detto una cazzata, qua sarebbe piuttosto complesso renderla effettivamente pura dato che c'è sia setRooms che setIsRoomsSearching, quindi boh va bene lasciarla così direi, se no finiamo per complicarla di più a sto giro